src/UCTSearch.cpp

```diff
@@ -121,22 +121,38 @@ void UCTSearch::update_root() {
 #endif
 }
 
+float UCTSearch::get_min_psa_ratio() const {
+    const auto mem_full = m_nodes / static_cast<float>(MAX_TREE_SIZE);
+    // If we are halfway through our memory budget, start trimming
+    // moves with very low policy priors.
+    if (mem_full > 0.5f) {
+        // Memory is almost exhausted, trim more aggressively.
+        if (mem_full > 0.95f) {
+            return 0.01f;
+        } else {
+            return 0.001f;
+        }
+    }
+    else {
+        return 0.0f;
+    }
+}
+
 SearchResult UCTSearch::play_simulation(GameState & currstate,
                                         UCTNode* const node) {
     const auto color = currstate.get_to_move();
     auto result = SearchResult{};
 
     node->virtual_loss();
 
-    if (!node->has_children()) {
+    if (node->expandable()) {
         if (currstate.get_passes() >= 2) {
             auto score = currstate.final_score();
             result = SearchResult::from_score(score);
         } else if (m_nodes < MAX_TREE_SIZE) {
-            auto mem_full_pct = m_nodes / static_cast<float>(MAX_TREE_SIZE);
             float eval;
             auto success =
-                node->create_children(m_nodes, currstate, eval, mem_full_pct);
+                node->create_children(m_nodes, currstate, eval, get_min_psa_ratio());
             if (success) {
```

Comment:
I think this may be breaking the search? You will return an eval result multiple times from the same node. I think we observed this makes the program weaker.